Disallow 64 byte transactions

Christewart opened the pull request.

mmgen committed:
Base58chk-encoded extended keys are always 111 characters long.  Amend wording of BIP accordingly.

-This results in a Base58-encoded string of up to 112 characters.
+This results in a Base58-encoded string of exactly 111 characters.

Version bytes: 0x0488b21e (“xpub”), 0x0488ade4 (“xprv”), 0x043587cf (“tpub”), 0x04358394 (“tprv”)

Largest “possible” key:
   kL = 0x0488b21effffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff (78 bytes)
   b58chk(kL) = xpubEPi3iGSX9RiyuXPTijevUmMctBDQs2TWCMgUd3qKp6qCgUc8RUsPdPBrRC6whFeWTg37DcmnJJiKFL73DH4sjdApJkXBD3vFcBP4xHq3fPY (111 chars)

Smallest “possible” key:
   kS = 0x043583940000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000 (78 bytes)
   b58chk(kS) = tprv8ZgxMBicQKsPcsbCVeqqF1KVdH7gwDJbxbzpCxDUsoXHdb6SnTPYxdwSAKDC6KKJzv7khnNWRAJQsRA8BBQyiSfYnRt6zuu4vZQGKjeW4YF (111 chars)

ysangkok committed:
Reject BIP-0060 (three years inactivity)

ysangkok committed:
Reject 199 (expired)

dplusplus1024 committed:
Update to include newer address types

Added bech32 and bech32m address types to reflect the newer SegWit and Taproot addresses.

Co-Authored-By: Reese Russell <reese.russell@ymail.com>

TheBlueMatt committed:
Copy BIP 21 into a new BIP 321 with only the header changed

jonatack committed:
BIP39: add license and copyright section

These are required per BIP-2, and they avoid user confusion as seen in

https://github.com/bitcoin/bips/pull/1395#issuecomment-2393915807

Christewart committed:
BIP53: Disallow 64-byte transactions

Co-authored-by: Mark "Murch" Erhardt <murch@murch.one>

Co-authored-by: Jon Atack <jon@atack.com>

nymius committed:
fix(BIP174,BIP375): typo in PSBT_OUT_SP_V0_LABEL

Assuming a by one increment in the keytype of the silent payments output
fields, the following numeral to 0x09 in the hexadecimal system is 0x0a,
not 0x10.

murchandamus committed:
Merge pull request #1680 from jonatack/2024-10-BIP39-license-and-copyright-section

BIP39: add license and copyright section

ajtowns commented:
cc @darosior

ajtowns commented on bip-XXXX.mediawiki:
"been enforced" --> "been prevented" ?

darosior commented:
huh, what?

ajtowns commented on bip-XXXX.mediawiki:
The 64-byte transaction can match to the concatenation of two neighbouring branches, not just two neighbouring leaves.

I think the "valid and invalid" bit is a little misleading -- if you want the hashes to match a valid 64B tx, then they have to give you a valid utxo as the prevout which requires the first hash to match 28 bytes (in counter and 27 bytes of the utxo's txid) which is cryptographically hard. That's also true for matching a coinbase tx (where the "txid" is all 0's).

Ending up with a plausible but invalid tx is the main attack here, in that you either convince a real node that a valid block is invalid, or you convince a light client that an invalid tx really exists and was confirmed.

Christewart replied:
>The 64-byte transaction can match to the concatenation of two neighbouring branches, not just two neighbouring leaves.

:+1:

>Ending up with a plausible but invalid tx is the main attack here, in that you either convince a real node that a valid block is invalid, or you convince a light client that an invalid tx really exists and was confirmed.

:+1:

ajtowns commented on bip-XXXX.mediawiki:
If you can do 224-bits worth of work, you can already break bitcoin's security assumptions anyway.

Christewart replied:
Agree, but its worth documenting IMO.

ajtowns commented on bip-XXXX.mediawiki:
I think the discussion around

https://delvingbitcoin.org/t/great-consensus-cleanup-revival/710/41

(re: efficiency of the merkle proof for light clients) is probably worth including here somewhere.

Christewart replied:
I was thinking about this yesterday, is there any risk of tapscript trees and this vuln? It seems like the same logic may be possible to fool a user into believing their Script was include in a tapscript tree? Maybe it isn't possible  since the entire Script would have to be exactly 64 bytes in size and the attacker doesn't get to choose any bytes to grind on?

@darosior Seems to be hinting at this in his post you cited as well

>Finally, it is important to not only consider lite clients. Proofs of inclusion of a transaction in a block are useful in other, existing or potential, applications. Examples include sidechains, or future Script changes which would allow to check a merkle proof directly in the interpreter.

Reorganizing the partial merkle tree format seems out of scope for this document IMO, but definitely food for thought.

ajtowns replied:
Because leaves are hashed with `hash_TapLeaf()` and branches are combined with `hash_TapBranch()` you can't mistake `hash(64-byte-leaf)` for `hash(32-byte-branch, 32-byte-branch)` with taproot/tapscript, without finding a sha256 collision.

jonatack committed:
Merge pull request #1754 from nymius/fix/typo-in-PSBT_OUT_SP_V0_LABEL-in-BIP-375-and-BIP-174

ajtowns committed:
Use code block instead of pre for markdown

jonatack committed:
Merge pull request #1759 from murchandamus/render-email-in-.md

Render author email addresses in markdown BIPs

achow101 committed:
328: Add test vectors

achow101 committed:
328: Add reference implementation

achow101 committed:
390: Add reference implementation

achow101 committed:
373: Add test vectors

achow101 committed:
373: Add reference implementation

achow101 committed:
328: Correct Created date

Date that the BIP number was assigned is 2024-06-04.

achow101 committed:
328: Draft -> Proposed

arminsabouri committed:
Fix typo in bip-0119

jonatack committed:
spelling: globally change "implementor" to "implementer"

Although the variant "implementor" predominated for much of the late 20th
century, today "implementer" is considered standard, and the former spelling
triggers the typos spelling checker.

jonatack committed:
Merge pull request #1765 from 0xBEEFCAF3/patch-2

BIP119: Fix `OP_EQUALVERIFY` typo

murchandamus committed:
Merge pull request #1766 from jonatack/2025-02-implementors-to-implementers

spelling: globally change "implementor" to "implementer"

murchandamus committed:
Merge pull request #1763 from achow101/390-ref-impl

390: Add reference implementation

murchandamus committed:
390: Fix created date

jonatack committed:
Merge pull request #1767 from murchandamus/390-created-date

390: Fix created date

achow101 committed:
373: Correct Created date

achow101 committed:
373: Draft -> Proposed

murchandamus committed:
Merge pull request #1764 from achow101/373-tests

373: test vectors, reference implementation, update to Proposed

murchandamus committed:
Merge pull request #1762 from achow101/328-tests

328: test vectors, reference implementation, update to Proposed

jonatack committed:
BIP159: emphasize minimum number of blocks

jonatack committed:
BIP159: clarify pruned means not signaling serving complete block chain

e.g. that NODE_NETWORK is not set

See reference implementation

https://github.com/bitcoin/bitcoin/pull/10387

and this comment in that pull

https://github.com/bitcoin/bitcoin/pull/10387#discussion_r156861038

jonatack committed:
BIP159: editorial fixups

achow101 committed:
373: Correct test data mismatches

Corrects the 2 mismatches in the test vectors pointed out in https://github.com/bitcoin/bips/pull/1764#issuecomment-2661667939

jonatack committed:
Merge pull request #1769 from achow101/373-test-fix

373: Correct test data mismatches

jonatack committed:
BIP159: unwillingly -> unwittingly

Not the same meaning, so not a purely editorial fixup, but I think "unwittingly"
expresses the intended meaning.

murchandamus committed:
BIP3: Update BIP Process

jonatack committed:
Merge pull request #1712 from murchandamus/2024-12-update-bip-process

BIP3: Updated BIP Process

murchandamus committed:
BIP3: Address follow-ups from #1712

costcould committed:
BIP0075: fix 404 status URL

Signed-off-by: costcould <fliter@myyahoo.com>

jonatack committed:
Merge pull request #1771 from murchandamus/2025-02-bip-3-follow-ups

BIP3: Address follow-ups from #1712

doc-hex committed:
BIP329: add optional data fields, fix a JSON type (#1750)

* Correct type error in sample JSON

* first pass: moar data

* more fiat and edits

* more edits

* add per-txn value

* Update bip-0329.mediawiki

Co-authored-by: Jon Atack <jon@atack.com>

* Update bip-0329.mediawiki

Co-authored-by: Jon Atack <jon@atack.com>

* Update bip-0329.mediawiki

Co-authored-by: Jon Atack <jon@atack.com>

* Update bip-0329.mediawiki

Co-authored-by: Jon Atack <jon@atack.com>

* Update bip-0329.mediawiki

Co-authored-by: Jon Atack <jon@atack.com>

* Update bip-0329.mediawiki

Co-authored-by: Jon Atack <jon@atack.com>

* Update bip-0329.mediawiki

Co-authored-by: Jon Atack <jon@atack.com>

* Update bip-0329.mediawiki

Co-authored-by: Jon Atack <jon@atack.com>

* edit

* include @craigraw feedback

* typo fix

* add comment

* further edits

* broaden uses for address.heights

---------

Co-authored-by: Jon Atack <jon@atack.com>

fjahr committed:
BIP 94: Move to Final

murchandamus committed:
Merge pull request #1774 from fjahr/bip94-status

BIP 94: Move status to Final

murchandamus committed:
BIP159 Risks section: clarifications and fixups

jonatack committed:
Merge pull request #1772 from costcould/master

BIP75: fix BIP70 extensions url and clarify text

craigraw committed:
add minor previously approved edits

craigraw committed:
add address heights clarification to align with transaction height specification

jonatack committed:
Merge pull request #1776 from craigraw/wallet-labels-edits

BIP329: PR 1750 follow-up edits

murchandamus committed:
Merge pull request #1768 from jonatack/2025-02-bip159

BIP159 updates

andrewtoth committed:
BIP374: Add message to rand computation (#1758)

* BIP374: Add message to rand computation

* BIP374: Update reference and test vectors

* Add changelog

* Format changelog according to BIP3

* Add creation date

Co-authored-by: Jon Atack <jon@atack.com>

* Grammar fix

Co-authored-by: Jon Atack <jon@atack.com>

* update changelog

---------

Co-authored-by: Jon Atack <jon@atack.com>

vipocenka committed:
fix gentestvectors.go

sky-coderay committed:
Update bip-0379.md

sky-coderay committed:
Update bip-0379.md

VolodymyrBg committed:
Implement subtraction operator for GE class in BIP-0374 reference code

This commit implements the subtraction operator (sub) for the GE (Group Element) class in the secp256k1.py file as requested in the TODO comment in reference.py.

The implementation is straightforward, leveraging the existing neg method to define subtraction as addition with the negated element: self + (-a).

After implementing the operator, the code in reference.py was simplified by replacing expressions like:
s * G + (-e * A) with s * G - e * A

This makes the code more readable and directly matches the mathematical notation used in the BIP-0374 specification.

Co-Authored-By: Sebastian Falbesoner <sebastian.falbesoner@gmail.com>

darosior committed:
Consensus Cleanup BIP draft

Roasbeef committed:
BIP-0094: add default p2p port for testnet4

Roasbeef committed:
BIP-0094: reformat specification section for clarity and readability

Restructured the specification section to make the consensus rules
clearer and more scannable. The previous section interleaved commentary
and historical tidbits with the motivation and new rules, making it
difficult to quickly identify the exact rule changes.

The updated format:
- Numbers each rule for easier reference
- Adds explicit "Rule Specification" sections
- Uses structured lists with MUST statements following RFC/IETF
  conventions
- Provides a clear problem statement before each solution
- Separates explanatory text from the actual rules

These changes make it much easier for implementers to understand what
changes are required without having to parse through multiple paragraphs
of text.

wgyt committed:
fix links in bip-0300

jonatack committed:
BIP119: appease typos linter

jonatack committed:
Merge pull request #1781 from Roasbeef/testnet4-port

BIP-0094: add default p2p port for testnet4

jonatack committed:
BIP94 PR1781 fixup

jonatack committed:
Merge pull request #1784 from jonatack/2025-03-appease-typos-linter-in-BIP119

BIP119: appease typos linter

jonatack committed:
Merge pull request #1785 from jonatack/2025-03-bip94-touchup

BIP94 PR1781 editorial fixups

EthnTuttle committed:
fix: grammatical typos

jonatack committed:
Merge pull request #1788 from EthnTuttle/patch-1

BIP119 fix grammatical typos

darosior committed:
bip-0348: correct reference to Tapscript unknown key types

Author confused bip-0341, which defines the Taproot construction, with bip-0342, which defines the
Tapscript scripting system. Unknown key types are defined in the latter, as part of the semantics of
the CHECKSIG{VERIFY} and CHECKSIGADD opcodes.

darosior committed:
Remove now unused reference to bip-0341

jonatack committed:
Merge pull request #1789 from darosior/2503_bip348_ref_bip342

bip-0348: correct BIP number in referencing unknown key types

jonatack committed:
Merge pull request #1783 from wgyt/wgyt-bips-fix-link

Fix links in BIP300

futreall committed:
Update bip-0375.mediawiki

JeremyRubin committed:
[BIP-119] language overhaul & cleanup

jonatack committed:
Merge pull request #1782 from Roasbeef/testnet4-spec-reword

BIP-0094: reformat specification section for clarity and readability

murchandamus committed:
Merge pull request #1790 from futreall/fix

bip-0375.mediawiki fix duplicate 'the'

murchandamus committed:
bip3: Move to Proposed

guspan-tanadi committed:
fix pdf path links bip-0068

murchandamus committed:
Merge pull request #1795 from guspan-tanadi/pdfpathlinks

bip-0068: current pdf path links

jonatack committed:
Merge pull request #1792 from JeremyRubin/119-edits-2025

[BIP-119] language overhaul & cleanup

murchandamus committed:
Merge branch 'BIP-0060'

jonatack committed:
Merge pull request #1777 from vipocenka/fix/fix

BIP158: minor error message fixup in `gentestvectors.go`

jonatack committed:
Merge pull request #1584 from jonatack/2024-04-bip32-amendment

BIP32 amendment: base58chk-encoded extended keys are always 111 chars

murchandamus committed:
Merge pull request #1778 from sky-coderay/patch-1

BIP379: add missing hexadecimal notations

wgyt committed:
fix links

murchandamus committed:
Merge pull request #1791 from wgyt/wgyt-bips-patch

BIPs 67 and 301: fix links

sky-coderay committed:
Update bip-0069_examples.py

jonatack committed:
Merge pull request #1798 from sky-coderay/master

BIP69: typo fixup in bip-0069_examples.py

jonatack committed:
Merge pull request #1794 from murchandamus/2025-03-propose-BIP3

BIP3: Move to Proposed

TheBlueMatt committed:
Update BIP 321 with information about more modern usage of it

As Bitcoin has grown, the introduction of new address formats
describing new forms of payment instructions has become
increasingly fraught with compatibility issues. Not only does there
exist traditional on-chain addresses, but some recipients wish to
receive Lightning (when the sender supports it) or newer formats
such as Silent Payments.

This has led to increasing use of the BIP 21 query parameters to
encode further optional payment instructions.

Looking forward, as new payment instructions get adopted, it makes
much more sense to include them in query parameters rather than
replace the existing address field, ensuring compatibility with
senders and recipients who may or may not be upgraded to support
all the latest payment instructions.

This updates BIP 321 to suggest that future address formats do this.

Further, it updates BIP 321 to allow an empty bitcoin address in
cases where new payment instructions have moved to becoming
mandatory. This isn't a backwards-incompatible change any more than
switching to a new address format is, so doesn't impact existing
BIP 21 implementations in a new way, however provides a nice
conclusion to the query-parameter-based upgrade path - once a form
of payment instructions has broad adoption, senders can simply drop
the existing address field, keeping their existing query parameter
encoding, rather than replace the existing address field. It also
addresses the question of what to do if a wallet no longer wishes
to receive some legacy on-chain address, but has multiple payment
instruction formats that they wish to include - deciding which one
to place in the address field would be a difficult task.

Finally, it defines a new query parameter, the `pop` parameter,
which allows the initiating application to receive callbacks for
proof of payment completion.

darosior committed:
bip3: rename 'shareholder' to 'stakeholder'

Shareholder refers to an individual or a legal entity owning a share of a company's share capital.
Since the Bitcoin system is not a company, but different actors across the industry have a stake in
its operation, i think the word "stakeholder" better conveys the intended meaning of the original
author here.

darosior committed:
bip3: link to ownership transfer section for complete->closed transition

Reading from top to bottom, the passive voice "they become BIP's author or deputy" left me wondering
how it would concretely work in practice. Link to the transferring ownership section for
clarification.

murchandamus committed:
Merge pull request #1804 from darosior/2503_bip3_nits

bip3: a couple nits

litvintech committed:
Update dead link in `bip-0003.md`

murchandamus committed:
BIP3: Fix link to BIP 123

quapka committed:
BIP380: minor grammar fixups (#1802)

* Use pronoun only after recalling the sentence object

* Use singular form

* Fix phrasal verb form

---------

Co-authored-by: quapka <>

real-or-random committed:
bip340: Change license of code and test vectors

See https://github.com/bitcoin/bips/commits/master/bip-0340 for a list
of contributors. I have obtained permission to do this change from all
contributors in private. Nevertheless, it would be good to get an ACK
from every contributor in order to have publicly available evidence.

 - [ ] @sipa
 - [ ] @jonasnick
 - [ ] @theStack
 - [ ] @ysangkok

I haven't contacted @Sajjon and @satsie, whose contributions constitute
of fixing not more than two typos and are thus below the threshold of
originality required for copyright to be applicable.

futreall committed:
fix error Base48 to Base58

chore: fix error Base48 to Base58

fix error Base48 to Base58

jonatack committed:
Merge pull request #1808 from futreall/master

BIP328: minor docstring fixups

jonatack committed:
Merge pull request #1807 from real-or-random/340-code-license-change

jonatack committed:
BIP340: fix url to test-vectors.py

murchandamus committed:
Merge pull request #1810 from jonatack/2025-04-bip340-url-fixup

BIP340: fix url to test-vectors.py

futreall committed:
fix err `has`  to  `hash` reference.py

murchandamus committed:
Merge pull request #1811 from futreall/master

BIP352: minor docstring fixups

leopardracer committed:
Update reference.py

leopardracer committed:
Update bip-0374.mediawiki

jonatack committed:
Merge pull request #1812 from leopardracer/master

BIP-0324: fix typo

murchandamus committed:
Merge pull request #1814 from leopardracer/patch-1

BIP-0374: fix typo

kilavvy committed:
BIP381, 382: minor grammar fixups (#1816)

* Update bip-0382.mediawiki

* Update bip-0381.mediawiki

BitcoinErrorLog committed:
Add BIP177: Redefine Bitcoin's Base Unit

- Redefine bitcoin base unit to smallest unit
- Propose BIP 21Q: Redefine bitcoin base unit to smallest indivisible unit
- Adds comments acknowledging and handling sats and satoshis
- Make use of "base unit" and variations more consistent and intentional
- Make "bitcoin" v "Bitcoin" consistent
- Made "bitcoin" v "Bitcoin" consistent by using Bitcoin for the protocol and idea, and bitcoin for the units, which I believe is conventional style.

zeevick10 committed:
BIP15,20: add missing words (#1822)

* Update bip-0015.mediawiki

* Update bip-0020.mediawiki

* Update bip-0035.mediawiki

leopardracer committed:
BIP75: fix typo (#1823)

wgyt committed:
BIP3,37,39,42,52,62: fix typos (#1824)


Co-authored-by: Mark "Murch" Erhardt <murch@murch.one>

VolodymyrBg committed:
BIPs 78 and 329: minor grammar and typo fix-ups (#1825)

* BIP 329 fix typo

* BIP 78 fix typos

jonatack committed:
Merge pull request #1779 from VolodymyrBg/AE5959595959

BIP374: add subtraction operator for GE class

Pronoss committed:
BIP-0374: fix incorrect bit index and modernize CSV reader usage in test vector scripts (#1817)

* Update run_test_vectors.py

* Update gen_test_vectors.py

* Regenerate test vectors after fixing message tampering logic in gen_test_vectors.py

jonatack committed:
Monthly typo fixups

Co-authored-by: xiaobei0715 <1505929057@qq.com>
Co-authored-by: wgyt <wgythe@gmail.com>
Co-authored-by: Ragnar <rodiondenmark@gmail.com>

NicolasDorier committed:
BIP388: List BTCPay Server and NBitcoin implementations

* BTCPay Server and NBitcoin supporting BIP388
* BIP388: Spelling improvements

Co-authored-by: Salvatore Ingala <6681844+bigspider@users.noreply.github.com>

murchandamus committed:
Merge pull request #1555 from TheBlueMatt/2024-03-uris-without-bodies

BIP 321: URI Scheme (Replace BIP 21 with a new BIP containing information about more modern usage of it)

ben-kaufman committed:
BIP48: Add p2tr script type derivation

gap-editor committed:
Update bip-0119.mediawiki

jonatack committed:
Merge pull request #1838 from gap-editor/bip-0119

BIP 119: fix link to MES16 paper

gap-editor committed:
bip-0112: fix links to Deployable Lightning paper (#1837)

0ceanSlim committed:
Add BIP172: Define Bitcoin Subunits as Satoshis

gap-editor committed:
BIP 154: fix link to cuckoo-profile.pdf (#1840)

* Update bip-0154.mediawiki

* Update bip-0154.mediawiki

gap-editor committed:
BIP 99: update invalid link to hardfork-timewarp-0.11 branch (#1836)

* Update bip-0099.mediawiki

* Update bip-0099.mediawiki

gap-editor committed:
Update bip-0197.mediawiki

murchandamus committed:
Merge pull request #1839 from jonatack/2025-04-collect-typo-fixups

Collect and curate typo fixups from April

murchandamus committed:
Merge pull request #1842 from gap-editor/bip-0197

BIP 197: Update the invalid link

jonatack committed:
Merge pull request #1800 from darosior/consensus_cleanup

BIP 54: Consensus Cleanup

jonatack committed:
Merge pull request #1835 from ben-kaufman/bip48-p2tr

BIP48: Add p2tr script type derivation

ben-kaufman committed:
BIP48: Add P2SH-P2WSH and P2TR mainnet examples

jamesob committed:
BIP-345: withdraw

murchandamus committed:
Merge pull request #1821 from BitcoinErrorLog/master

BIP177: Redefine Bitcoin’s Base Unit

murchandamus committed:
Merge pull request #1841 from 0ceanSlim

- Adds BIP172: Define Bitcoin Subunits as Satoshis

jonatack committed:
Merge pull request #1846 from ben-kaufman/bip48-mainnet-examples

BIP48: Add P2SH-P2WSH and P2TR mainnet examples

murchandamus committed:
Merge pull request #1848 via 'jamesob-25-05-withdraw-vault'
